test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,443 @@
+			targetBlockDev, fmt.Sprintf("%s is not a directory", targetBlockDev), &hostPathDirectory)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathFile", func() {
+		verifyPodHostPathTypeFailure(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName},
+			targetBlockDev, fmt.Sprintf("%s is not a file", targetBlockDev), &hostPathFile)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathSocket", func() {
+		verifyPodHostPathTypeFailure(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName},
+			targetBlockDev, fmt.Sprintf("%s is not a socket", targetBlockDev), &hostPathSocket)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathCharDev", func() {
+		verifyPodHostPathTypeFailure(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName},
+			targetBlockDev, fmt.Sprintf("%s is not a character device", targetBlockDev), &hostPathCharDev)
+	})
+})
+
+func newHostPathTypeTestPod(nodeSelector map[string]string, hostDir, mountDir string, hostPathType *v1.HostPathType) *v1.Pod {
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-hostpath-type-",
+		},
+		Spec: v1.PodSpec{
+			NodeSelector:  nodeSelector,
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				{
+					Name:  "host-path-testing",
+					Image: imageutils.GetE2EImage(imageutils.Agnhost),
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "host",
+							MountPath: mountDir,
+							ReadOnly:  false,
+						},
+					},
+				},
+			},
+			Volumes: []v1.Volume{
+				{
+					Name: "host",
+					VolumeSource: v1.VolumeSource{
+						HostPath: &v1.HostPathVolumeSource{
+							Path: hostDir,
+							Type: hostPathType,
+						},
+					},
+				},
+			},
+		},
+	}
+	return pod
+}
+
+func verifyPodHostPathTypeFailure(f *framework.Framework, nodeSelector map[string]string, hostDir, pattern string, hostPathType *v1.HostPathType) {
+	pod := newHostPathTypeTestPod(nodeSelector, hostDir, "/mnt/test", hostPathType)
+	ginkgo.By(fmt.Sprintf("Creating pod %s", pod.Name))
+	pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(context.TODO(), pod, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+
+	ginkgo.By("Checking for HostPathType error event")
+	eventSelector := fields.Set{
+		"involvedObject.kind":      "Pod",
+		"involvedObject.name":      pod.Name,
+		"involvedObject.namespace": f.Namespace.Name,
+		"reason":                   events.FailedMountVolume,
+	}.AsSelector().String()
+	msg := "hostPath type check failed"
+
+	err = e2eevents.WaitTimeoutForEvent(f.ClientSet, f.Namespace.Name, eventSelector, msg, framework.PodStartTimeout)
+	// Events are unreliable, don't depend on the event. It's used only to speed up the test.
+	if err != nil {
+		framework.Logf("Warning: did not get event about FailedMountVolume")
+	}
+
```

Comment:
Ah sorry one more check to make sure the pod is not running. Like https://github.com/kubernetes/kubernetes/blob/b9c370c90f71d4a6a5ad5d4668de6d3fb987720c/test/e2e/storage/testsuites/volumemode.go#L245